### PR TITLE
nixos/stalwart-mail: add openFirewall option

### PR DESCRIPTION
**Pull request by onny**
## Description of changes

- add option `openFirewall` which reads all ports of the `settings.server.listener` attribute set and appends it to `networking.firewall.allowedTCPPorts`

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by pacien — COMMENTED** [review]
I don't think that we should configure listeners implicitly by default.
Protocol choices are, more often than not, quite custom for email servers.

In particular, I'd expect users of stalwart-mail to be using it for its JMAP
implementation instead of IMAP and SMTP submission. Not to mention the
listeners for Stalwart's admin API, web UI, manage sieve, LMTP, POP, and so on,
which may or may not be desirable depending on the setup.

(Also, if I recall correctly, the proposed default with TLS SMTP on port 465
is considered deprecated in favour of SMTP submission with STARTTLS on port
587).

I'd therefore prefer starting with a blank slate regarding the listeners,
allowing one to pick and choose to their explicit needs.

This goes together with the firewall configuration.
Opening all listener ports to the world does not seem to be fitting,
in particular for the admin endpoint or LMTP, that one should restrict
to local access, or the web and JMAP ports which are quite likely to stay
behind a reverse proxy like nginx.

Also, the proposed implementation seems to open firewall ports on all
interfaces, even when a listener is defined on a particular one. This seems
complicated to implement automatically with more granularity. So I think
it should also be left to the user to define manually here too.

**Comment by onny** [comment]
@pacien Agree on that 👍
`openFirewall` is used by most modules to open TCP ports on all interfaces. I would like to use this since my mail server should be open to the public.
Maybe I could add a note to the description what the option exactly does?
Manually adding the listener ports to allowedTCPPorts is a bit confusing and unclear.

File: nixos/modules/services/mail/stalwart-mail.nix [hunk 1]
```diff
@@ -9,22 +9,31 @@ let
...
-
-        By default, the module is configured to store everything locally.
+        Whether to open TCP firewall ports, which are specified in
+        {option}`services.stalwart-mail.settings.listener` on all interfaces.
       '';
     };
+
+    settings = mkOption {
+      inherit (configFormat) type;
+    };
```
[the remark this answers is not in the document]
Comment (onny) [reply nixos/modules/services/mail/stalwart-mail.nix]: Thanks fixed :)

**Review by pacien — APPROVED** [review]
Looks good to me 👍

**Pull request merged**